File: ViewSysterm/MyCustomView/app/src/main/java/com/wyj/view/timeline/MainActivity.java
package com.wyj.view.timeline;

import android.graphics.Color;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import com.wyj.view.R;

public class MainActivity extends AppCompatActivity {
    private RecyclerView Rv;
    private LogisticsAdapter myAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initData();
    }

    // 初始化显示的数据
    public void initData() {
        List<LogisticsInfo> logisticsInfos = new ArrayList<>();
        LogisticsInfo logisticsInfo = new LogisticsInfo();
        logisticsInfo.setState("已签收");
        logisticsInfo.setDateStr("06-12 11:07");
        logisticsInfo.setDesc("包裹已签收！包裹已签收！（凭取件码）如有问题请电联：代收点15827065070/何林峰 17366929255，" +
                "投诉电话：027-87443612 在线求五星 点亮兔兔小行星，只为服务您满意~");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setState("代取件");
        logisticsInfo.setDateStr("06-11 12:18");
        logisticsInfo.setDesc("您的包裹已放置【代收点】，记得早点来【花城南路42附51号申通快递】取它回家！如有问题请联系：代收点" +
                " 15827065070/何林峰 17366929255，投书电话：027-87443612");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setState("派送中");
        logisticsInfo.setDateStr("06-11 08:03");
        logisticsInfo.setDesc("【武汉光谷生物城网点】的兔兔快递员 何林峰 正在派件（可放心接听952300专属热线），" +
                "投诉电话：027-87443612。今天的兔兔，怀揣包裹，卯足干劲，为您“加吉”派送中");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setState("运送中");
        logisticsInfo.setDateStr("06-11 08:03");
        logisticsInfo.setDesc("【快件到达【武汉光谷生物城网点】");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setDateStr("06-11 03:29");
        logisticsInfo.setDesc("快件离开【武汉转运中心B1】已发往【武汉光谷生物城网点】");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setDateStr("06-11 03:03");
        logisticsInfo.setDesc("快件到达【武汉转运中心B1】");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setDateStr("06-10 02:46");
        logisticsInfo.setDesc("快件离开【揭阳转运中心】已发往【武汉转运中心B1】");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setDateStr("06-10 00:58");
        logisticsInfo.setDesc("快件离开【潮安集散点】已发往【揭阳转运中心】");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setDateStr("06-09 20:21");
        logisticsInfo.setDesc("快件离开【潮州彩塘网点】已发往【潮安集散点】");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setState("已揽件");
        logisticsInfo.setDateStr("06-09 20:11");
        logisticsInfo.setDesc("快件离开【潮州彩塘网点】已发往【潮安集散点】");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setState("已发货");
        logisticsInfo.setDateStr("06-09 17:37");
        logisticsInfo.setDesc("包裹正在等待揽收");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setState("已出库");
        logisticsInfo.setDateStr("06-09 17:37");
        logisticsInfo.setDesc("您的包裹已出库");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setState("已下单");
        logisticsInfo.setDateStr("06-09 17:37");
        logisticsInfo.setDesc("您的包裹已打包");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setDateStr("06-09 17:36");
        logisticsInfo.setDesc("您的订单已打物流单");
        logisticsInfos.add(logisticsInfo);

        logisticsInfo = new LogisticsInfo();
        logisticsInfo.setDateStr("06-09 17:27");
        logisticsInfo.setDesc("商品已经下单");
        logisticsInfos.add(logisticsInfo);

        Rv = (RecyclerView) findViewById(R.id.my_recycler_view);
        //使用线性布局
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        Rv.setLayoutManager(layoutManager);
        Rv.setHasFixedSize(true);

        //用自定义分割线类设置分割线
        LogisticsItemDecoration itemDecoration = new LogisticsItemDecoration.Builder()
                .setDot(12, Color.parseColor("#FF555555"), 8)
                .setLine(2, 40, Color.parseColor("#44222222"))
                .setOffsets(58, 50)
                .build();
        Rv.addItemDecoration(itemDecoration);

        myAdapter = new LogisticsAdapter(this, logisticsInfos);
        Rv.setAdapter(myAdapter);
    }

}

File: ImageCache/app/src/main/java/com/wyj/imagecache/MainActivity.kt
package com.wyj.imagecache

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.util.Log
import android.util.LruCache
import android.widget.Button
import android.widget.ImageView
import okhttp3.*
import java.io.IOException
import java.lang.ref.WeakReference

/**
 *  https://juejin.cn/post/6844903855818276871
 */
class MainActivity : AppCompatActivity() {
    companion object {
        private const val TAG = "MainActivity"
        private const val SUCCESS = 0x0001
        private const val FAIL = 0x0002
    }

    private val imageUrl = "https://cn.bing.com/sa/simg/hpb/LaDigue_EN-CA1115245085_1920x1080.jpg"
    private var imageLoader: MyImageLoader? = null
    private var iv: ImageView? = null
    private var handler: MyHandler? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        imageLoader = MyImageLoader()
        handler = MyHandler(this)
        iv = findViewById<ImageView>(R.id.iv_lrucache)
        findViewById<Button>(R.id.bt_load).setOnClickListener {
            imageLoader?.getBitmap(imageUrl)?.let {
                Log.d(TAG, "onCreate: wyj has memory cache")
                iv?.setImageBitmap(it)
            } ?: downloadBitmap()
        }
    }

    private fun downloadBitmap() {
        Log.d(TAG, "downloadBitmap: wyj 从网络下载图片")
        val okHttpClient = OkHttpClient()
        val request: Request = Request.Builder()
            .url(imageUrl)
            .build()
        val call = okHttpClient.newCall(request)
        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                Log.d(TAG, "onFailure: wyj download image failed.")
                handler?.obtainMessage(FAIL)?.let { handler?.sendMessage(it) }
            }

            override fun onResponse(call: Call, response: Response) {
                val bytes = response.body()?.bytes()
                val message = handler?.obtainMessage()?.apply {
                    what = SUCCESS
                    obj = bytes
                    handler?.sendMessage(this)
                }
            }

        })
    }

    private fun onDownloaded(bytes: ByteArray) {
        val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
        imageLoader?.addBitmap(imageUrl, bitmap)
        iv?.setImageBitmap(bitmap)
    }

    @SuppressLint("HandlerLeak")
    class MyHandler(activity: MainActivity): Handler() {
        private var reference: WeakReference<MainActivity>
        init {
            reference = WeakReference(activity)
        }

        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            when(msg.what) {
                SUCCESS -> {
                    val activity = reference.get()
                    activity?.let {
                        if (!it.isFinishing) {
                            if (msg.obj is ByteArray) {
                                it.onDownloaded(msg.obj as ByteArray)
                            }
                        }
                    }
                }
                FAIL -> {
                    Log.d(TAG, "handleMessage: wyj 图片下载失败")
                }
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        handler?.let {
            it.removeCallbacksAndMessages(null)
            handler = null
        }
    }
}